src/ronin/basefee-treasury/RoninBaseFeeTreasury.sol

```diff
@@ -0,0 +1,546 @@
+
+    for (uint256 i; i < length; ++i) {
+      if (getState(allHashes[i]) == State.Active) {
+        pendingHashes[count] = allHashes[i];
+        ++count;
+      }
+    }
+
+    assembly ("memory-safe") {
+      mstore(pendingHashes, count)
+    }
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function getSnapshotVotePowerById(bytes32 hash, address cid) external view returns (uint256 votePower) {
+    return _info[hash]._votePower[cid];
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function getMinimumVotePowerToPass(
+    bytes32 hash
+  ) external view returns (uint256 votePower) {
+    return _calcMinVoteFor(_info[hash]._totalVotePow);
+  }
+
+  /**
+   * @inheritdoc IBaseFeeTreasury
+   */
+  function getMinimumVotePowerToCancel(
+    bytes32 hash
+  ) external view returns (uint256 minVoteAgainst) {
+    return _calcMinVoteAgainst(_info[hash]._totalVotePow);
+  }
+
+  /**
+   * @dev Create a new proposal.
+   */
+  function _createProposal(bytes32 hash, Proposal memory p, uint256 totalVotePower) internal {
+    ProposalInfo storage $ = _info[hash];
+
+    $._createdAt = uint40(block.timestamp);
+    $._state = State.Active;
+    $._proposal = p;
+    $._totalVotePow = (totalVotePower).toUint96();
+
+    require(_pendingProposals.add(hash), ErrExistedProposal(hash));
+
+    emit ProposalCreated(msg.sender, p.proposer, hash, p, $._createdAt);
+  }
+
+  /**
+   * @dev Snapshot the vote power of all candidates at the time of proposal creation.
+   */
+  function _snapshotVotePower(
+    bytes32 hash
+  ) internal returns (uint256 totalVotePower) {
+    address[] memory allCids = ICandidateManager(getContract(ContractType.VALIDATOR)).getValidatorCandidateIds();
+    uint256[] memory selfStakes = IStaking(getContract(ContractType.STAKING)).getManySelfStakingsById(allCids);
+
+    uint256 length = allCids.length;
+    require(length == selfStakes.length, ErrLengthMismatch(msg.sig));
+
+    ProposalInfo storage $ = _info[hash];
+
+    for (uint256 i; i < length; ++i) {
+      uint96 scaledStake = (selfStakes[i] / 1 ether).toUint96();
+      totalVotePower += scaledStake;
+
+      $._votePower[allCids[i]] = scaledStake;
+
+      // Self-stake is expected to > 1e18
+      require(scaledStake != 0, ErrPanicDataCorruption(allCids[i]));
+    }
+  }
+
+  /**
+   * @dev Vote on a proposal.
+   */
+  function _vote(bytes32 hash, address cid, Vote v) internal {
+    ProposalInfo storage $ = _info[hash];
+    IProfile profile = IProfile(getContract(ContractType.PROFILE));
+
+    _requireValidNonce($._proposal.nonce);
+    require(v == Vote.For || v == Vote.Against, ErrInvalidVote());
+    require($._votePower[cid] != 0, ErrNewlyRegisteredOrRenouncedOrEmergencyExit());
+    require($._vote[cid] == Vote.Unknown, ErrAlreadyVoted(hash, cid, $._vote[cid]));
+    require(
+      _computePeriod(profile.getId2RegisteredAt(cid) + COOLDOWN_PERIOD) <= _computePeriod($._createdAt),
+      ErrNewlyRegisteredCannotVote(cid)
+    );
```

Comment:
This condition prevents new candidates from voting on past proposals. Is it as expected? 
Even if someone registers as a candidate and waits a day, he still can’t vote on past proposals.

Reply:
As expected.